### PR TITLE
Apply a few recent chages in Quat and Basis to their respective Mono …

### PR DESCRIPTION
**Pull request by ghost**
…counterparts.

### CODE REVIEW

**Comment by tagcup** [comment]
I was trying to run the attached project in #11901 locally (with a godot including the modification in this PR), but on the first run, it crashes with the following:

```
Unhandled Exception:
System.TypeLoadException: Could not resolve type with token 01000036 (from typeref, class/assembly System.Linq.Enumerable, System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
  at Godot.GodotTaskScheduler.Activate () [0x0000b] in <18bb9f788af2492cb13178a44947a28f>:0 
  at (wrapper native-to-managed) Godot.GodotTaskScheduler:Activate (Godot.GodotTaskScheduler,System.Exception&)
ERROR: frame: Method/Function Failed.
   At: modules/mono/csharp_script.cpp:331.

Unhandled Exception:
System.TypeLoadException: Could not resolve type with token 01000036 (from typeref, class/assembly System.Linq.Enumerable, System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Could not resolve type with token 01000036 (from typeref, class/assembly System.Linq.Enumerable, System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
```

After restarting and trying to run it again, I got this error from Mono:

`Core/Basis(187,20): Use of unassigned local variable  "euler"`. Not much familiar with C#, but I tried changing the definition of euler to `Vector3 euler(0f,0f,0f)` (it doesn't have a "default" constructor apparently), and also added lines like euler.z = 0, such that `return euler` was no longer on line 187.

I recompiled godot, and tried to run the project, but I'm getting exactly the same error from Mono, at exactly the same line, which is weird. Should I open an issue for this?

**Comment by tagcup** [comment]
Tried creating a new project from scratch, when I press create C# Solution, it fails and I get this error:

```
Unhandled Exception:
System.TypeLoadException: Failure has occurred while loading a type.
ERROR: generate_game_project: Method/Function Failed, returning: String()
   At: modules/mono/editor/csharp_project.cpp:96.
from file: res://default_env.tres
from file: res://Node.tscn
cached: res://default_env.tres
cached: res://Node.tscn

Unhandled Exception:
System.TypeLoadException: Failure has occurred while loading a type.
ERROR: generate_game_project: Method/Function Failed, returning: String()
   At: modules/mono/editor/csharp_project.cpp:96.

```

**Comment by tagcup** [comment]
This is all with the current master 5a49106 + the changes in this PR

**Comment by neikeq** [comment]
Yes, there is something wrong there. I'll check.

**Comment by neikeq** [comment]
Fixed that bug with #11924

**Comment by tagcup** [comment]
Yeah, it doesn't crash or give any error message.
New changes I'm making in Basis.cs still don't affect new builds though. Is there something special that I'm supposed to do when I make a change in Basis.cs etc?

**Comment by neikeq** [comment]
@tagcup Since the API hashes didn't change, the C# bindings won't be regenerated. If you delete the `user://mono/solutions` folder, it will regenerate the bindings and build the API assemblies again.

**Comment by tagcup** [comment]
Ah, OK. I removed ~/.godot/mono and it worked. Do you think it should be automated, maybe by checking a hash or something?

BTW I updated the PR which seems to be working OK now.

**Comment by akien-mga** [comment]
Oops I used "Squash and merge" instead of "Create a merge commit" by mistake, sorry about that.

**Comment by neikeq** [comment]
I didn't finish review. I was comparing it to my local changes to Basis :(

File: modules/mono/glue/cs_files/Basis.cs [hunk 1]
```diff
@@ -347,6 +350,48 @@ public Vector3 xform_inv(Vector3 v)
...
+				float inv_s = 1f / s;
+				return new Quat(
+					(x[2] + z[0]) * inv_s,
+					(y[2] + z[1]) * inv_s,
+					s * 0.25f,
+					(y[0] - x[1]) * inv_s
+				);
+			}
+		}
+
```
Comment (neikeq) [inline modules/mono/glue/cs_files/Basis.cs]: What is this for? I don't see it anywhere in the Godot API?

File: modules/mono/glue/cs_files/Transform.cs [hunk 2]
```diff
@@ -104,6 +104,12 @@ public Transform(Vector3 xAxis, Vector3 yAxis, Vector3 zAxis, Vector3 origin)
             this.origin = origin;
         }
 
+        public Transform(Quat quat, Vector3 origin)
+        {
+            this.basis = new Basis(quat);
+            this.origin = origin;
+        }
+
```
Comment (neikeq) [inline modules/mono/glue/cs_files/Transform.cs]: I don't see this anywhere in the Godot API either.

File: modules/mono/glue/cs_files/Basis.cs (hunk 1 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: That's how you convert an orthogonal matrix to quaternion. If it's not there it's a bug.

File: modules/mono/glue/cs_files/Transform.cs (hunk 2 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Transform.cs]: That also would be need to be added then. It's there in GDScript

Comment (neikeq) [reply modules/mono/glue/cs_files/Transform.cs]: I don't see think constructor in GDScript.

File: modules/mono/glue/cs_files/Basis.cs (hunk 1 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: If you mean the function itself it's operator Quat function in Basis. Can be exposed better for gdnative I guess

File: modules/mono/glue/cs_files/Transform.cs (hunk 2 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Transform.cs]: But there is

Comment (tagcup) [reply modules/mono/glue/cs_files/Transform.cs]: If there are other missing constructors from Basis in Transform those should also be added in core/math so don't worry I'll add any missing ones

File: modules/mono/glue/cs_files/Basis.cs (hunk 1 above)
Comment (neikeq) [reply modules/mono/glue/cs_files/Basis.cs]: But this is not exposed to scripts. We can't diverge like this from Godot's API. In the case we actually want to expose something different, it should probably be via extension methods (or making the type partial and separating it in multiple files, to make it easy to mantain).

File: modules/mono/glue/cs_files/Transform.cs (hunk 2 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Transform.cs]: Of course they'll have origin as the extra parameter

File: modules/mono/glue/cs_files/Basis.cs (hunk 1 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Basis.cs]: Again don't worry. I'll add those. They are supposed to be there

File: modules/mono/glue/cs_files/Transform.cs (hunk 2 above)
Comment (tagcup) [reply modules/mono/glue/cs_files/Transform.cs]: So in GDScript, Transform(quat, origin) does work but when I checked the documentation, the closest constructor listed there is Transform(Basis, Vector3). I don't know if this is a documentation mistake or not, but if it's not, it seems GDScript can automatically convert a Basis to Quat?

That constructor should be added anyway though, I'm just surprised GDScript could do that

**Pull request merged**